File: Src/java/engine-fhir/src/test/java/org/opencds/cqf/cql/engine/fhir/data/EvaluatedResourcesTest.java
package org.opencds.cqf.cql.engine.fhir.data;

import static org.opencds.cqf.cql.engine.fhir.data.EvaluatedResourceTestUtils.CONDITION;
import static org.opencds.cqf.cql.engine.fhir.data.EvaluatedResourceTestUtils.ENCOUNTER;
import static org.opencds.cqf.cql.engine.fhir.data.EvaluatedResourceTestUtils.RETRIEVE_PROVIDER;
import static org.opencds.cqf.cql.engine.fhir.data.EvaluatedResourceTestUtils.assertEvaluationResult;

import java.util.List;
import java.util.Set;
import javax.annotation.Nonnull;
import org.junit.jupiter.api.Test;
import org.opencds.cqf.cql.engine.data.CompositeDataProvider;
import org.opencds.cqf.cql.engine.execution.CqlEngine;

class EvaluatedResourcesTest extends FhirExecutionTestBase {

    @Test
    void withCache() {
        CqlEngine engine = getCqlEngineForFhir(true);

        var results = engine.evaluate(library.getIdentifier(), Set.of("Union"));

        assertEvaluationResult(results, "Union", List.of(CONDITION, ENCOUNTER));

        results = engine.evaluate(library.getIdentifier(), Set.of("Encounter"));
        assertEvaluationResult(results, "Encounter", List.of(ENCOUNTER));

        results = engine.evaluate(library.getIdentifier(), Set.of("Condition"));
        assertEvaluationResult(results, "Condition", List.of(CONDITION));
    }

    @Test
    void withoutCache() {
        CqlEngine engine = getCqlEngineForFhir(false);

        var results = engine.evaluate(library.getIdentifier(), Set.of("Union"));

        assertEvaluationResult(results, "Union", List.of(CONDITION, ENCOUNTER));

        results = engine.evaluate(library.getIdentifier(), Set.of("Encounter"));
        assertEvaluationResult(results, "Encounter", List.of(ENCOUNTER));

        results = engine.evaluate(library.getIdentifier(), Set.of("Condition"));
        assertEvaluationResult(results, "Condition", List.of(CONDITION));

        results = engine.evaluate(library.getIdentifier(), Set.of("Union"));
        assertEvaluationResult(results, "Union", List.of(CONDITION, ENCOUNTER));
    }

    @Nonnull
    private CqlEngine getCqlEngineForFhir(boolean expressionCaching) {
        CqlEngine engine = getEngine();
        engine.getState()
                .getEnvironment()
                .registerDataProvider(
                        "http://hl7.org/fhir", new CompositeDataProvider(r4ModelResolver, RETRIEVE_PROVIDER));
        engine.getCache().setExpressionCaching(expressionCaching);
        return engine;
    }
}

File: Src/java/engine-fhir/src/test/kotlin/org/opencds/cqf/cql/engine/fhir/data/EvaluatedResourcesTest.kt
package org.opencds.cqf.cql.engine.fhir.data

import org.junit.jupiter.api.Test
import org.opencds.cqf.cql.engine.data.CompositeDataProvider
import org.opencds.cqf.cql.engine.execution.CqlEngine

internal class EvaluatedResourcesTest : FhirExecutionTestBase() {
    @Test
    fun withCache() {
        val engine = getCqlEngineForFhir(true)

        var results =
            engine
                .evaluate {
                    this.library(library!!.identifier!!)
                    expression("Union")
                }
                .onlyResultOrThrow

        EvaluatedResourceTestUtils.assertEvaluationResult(
            results,
            "Union",
            listOf(EvaluatedResourceTestUtils.CONDITION, EvaluatedResourceTestUtils.ENCOUNTER),
        )

        results =
            engine
                .evaluate {
                    this.library(library!!.identifier!!)
                    expression("Encounter")
                }
                .onlyResultOrThrow
        EvaluatedResourceTestUtils.assertEvaluationResult(
            results,
            "Encounter",
            listOf(EvaluatedResourceTestUtils.ENCOUNTER),
        )

        results =
            engine
                .evaluate {
                    this.library(library!!.identifier!!)
                    expression("Condition")
                }
                .onlyResultOrThrow
        EvaluatedResourceTestUtils.assertEvaluationResult(
            results,
            "Condition",
            listOf(EvaluatedResourceTestUtils.CONDITION),
        )
    }

    @Test
    fun withoutCache() {
        val engine = getCqlEngineForFhir(false)

        var results =
            engine
                .evaluate {
                    this.library(library!!.identifier!!)
                    expression("Union")
                }
                .onlyResultOrThrow

        EvaluatedResourceTestUtils.assertEvaluationResult(
            results,
            "Union",
            listOf(EvaluatedResourceTestUtils.CONDITION, EvaluatedResourceTestUtils.ENCOUNTER),
        )

        results =
            engine
                .evaluate {
                    this.library(library!!.identifier!!)
                    expression("Encounter")
                }
                .onlyResultOrThrow
        EvaluatedResourceTestUtils.assertEvaluationResult(
            results,
            "Encounter",
            listOf(EvaluatedResourceTestUtils.ENCOUNTER),
        )

        results =
            engine
                .evaluate {
                    this.library(library!!.identifier!!)
                    expression("Condition")
                }
                .onlyResultOrThrow
        EvaluatedResourceTestUtils.assertEvaluationResult(
            results,
            "Condition",
            listOf(EvaluatedResourceTestUtils.CONDITION),
        )

        results =
            engine
                .evaluate {
                    this.library(library!!.identifier!!)
                    expression("Union")
                }
                .onlyResultOrThrow
        EvaluatedResourceTestUtils.assertEvaluationResult(
            results,
            "Union",
            listOf(EvaluatedResourceTestUtils.CONDITION, EvaluatedResourceTestUtils.ENCOUNTER),
        )
    }

    private fun getCqlEngineForFhir(expressionCaching: Boolean): CqlEngine {
        engine.state.environment.registerDataProvider(
            "http://hl7.org/fhir",
            CompositeDataProvider(r4ModelResolver!!, EvaluatedResourceTestUtils.RETRIEVE_PROVIDER),
        )
        engine.cache.setExpressionCaching(expressionCaching)
        return engine
    }
}
